File: src/shared/lib/cache.ts
type CacheEntry<T> = { value: T; expiresAt: number }

export interface Cache<T> {
    get(key: string): T | null
    set(key: string, value: T, ttlMs?: number): void
    delete(key: string): void
    has(key: string): boolean
    size(): number
}

export class LRUCache<T> implements Cache<T> {
    private store: Map<string, CacheEntry<T>>
    private capacity: number
    private defaultTTL: number

    constructor(capacity: number, defaultTTL: number) {
        this.store = new Map()
        this.capacity = Math.max(1, capacity)
        this.defaultTTL = Math.max(1, defaultTTL)
    }

    get(key: string): T | null {
        const now = Date.now()
        const entry = this.store.get(key)
        if (!entry) return null
        if (now > entry.expiresAt) {
            this.store.delete(key)
            return null
        }
        this.touch(key, entry)
        return entry.value
    }

    set(key: string, value: T, ttlMs?: number): void {
        const expiresAt = Date.now() + (ttlMs ?? this.defaultTTL)
        const entry: CacheEntry<T> = { value, expiresAt }
        if (this.store.has(key)) {
            this.store.delete(key)
        }
        this.store.set(key, entry)
        this.evictIfNeeded()
    }

    delete(key: string): void {
        this.store.delete(key)
    }

    has(key: string): boolean {
        const now = Date.now()
        const entry = this.store.get(key)
        if (!entry) return false
        if (now > entry.expiresAt) {
            this.store.delete(key)
            return false
        }
        return true
    }

    size(): number {
        return this.store.size
    }

    private evictIfNeeded() {
        while (this.store.size > this.capacity) {
            const oldestKey = this.store.keys().next().value as string
            this.store.delete(oldestKey)
        }
    }

    private touch(key: string, entry: CacheEntry<T>) {
        this.store.delete(key)
        this.store.set(key, entry)
    }
}

export const defaultCache: Cache<any> = new LRUCache<any>(512, 60_000)

export function cacheGet<T>(key: string): T | null {
    return defaultCache.get(key) as T | null
}

export function cacheSet<T>(key: string, value: T, ttlMs: number): void {
    defaultCache.set(key, value, ttlMs)
}

export function makeKey(parts: Array<string | number | undefined>): string {
    return parts.filter((x) => x !== undefined).join('|')
}

export function invalidateByPrefix(prefix: string) {
    // naive prefix invalidation for in-memory cache
    const keys = Array.from((defaultCache as any).store?.keys?.() || []) as string[]
    for (const k of keys) {
        if (k.startsWith(prefix)) (defaultCache as any).delete(k)
    }
}

File: pyapp/src/pyapp/shared/cache.py
"""轻量级基于内存的 LRU 缓存实现。

提供带 TTL 的键值缓存，支持容量控制与最近使用更新，满足接口层去重与热点响应加速。
"""

from typing import Any, Dict, Optional
import time

class CacheEntry:
    """缓存条目，包含值与过期时间。"""

    def __init__(self, value: Any, ttl_seconds: int) -> None:
        self.value = value
        self.expires_at = time.time() + ttl_seconds

class LRUCache:
    """简易 LRU 缓存。

    参数：
        capacity: 最大保留键数。
        default_ttl_seconds: 默认过期秒数。
    """

    def __init__(self, capacity: int = 512, default_ttl_seconds: int = 60) -> None:
        self._store: Dict[str, CacheEntry] = {}
        from typing import List
        self._order: List[str] = []
        self._capacity = max(1, capacity)
        self._default_ttl = max(1, default_ttl_seconds)

    def _touch(self, key: str) -> None:
        try:
            self._order.remove(key)
        except ValueError:
            pass
        self._order.append(key)

    def _evict_if_needed(self) -> None:
        while len(self._order) > self._capacity:
            oldest = self._order.pop(0)
            self._store.pop(oldest, None)

    def get(self, key: str) -> Optional[Any]:
        """获取键的缓存值，过期则返回 `None` 并清理。"""
        entry = self._store.get(key)
        if not entry:
            return None
        if entry.expires_at < time.time():
            self._store.pop(key, None)
            try:
                self._order.remove(key)
            except ValueError:
                pass
            return None
        self._touch(key)
        return entry.value

    def set(self, key: str, value: Any, ttl_seconds: Optional[int] = None) -> None:
        """写入键值并更新最近使用顺序，必要时执行淘汰。"""
        ttl = ttl_seconds if ttl_seconds is not None else self._default_ttl
        self._store[key] = CacheEntry(value, ttl)
        self._touch(key)
        self._evict_if_needed()

    def delete(self, key: str) -> None:
        """删除指定键的缓存记录。"""
        self._store.pop(key, None)
        try:
            self._order.remove(key)
        except ValueError:
            pass

cache = LRUCache()
